File: backend/src/domain/value-objects/impression.ts
export class Impression {
  private readonly _value: string | undefined;

  private constructor(value: string | undefined | null) {
    if (value === null || value === undefined || value.trim() === '') {
      this._value = undefined;
    } else {
      const trimmed = value.trim();
      if (trimmed.length > 191) {
        throw new Error('Impression must be 191 characters or less');
      }
      this._value = trimmed;
    }
  }

  static create(value: string | undefined | null): Impression {
    return new Impression(value);
  }

  get value(): string | undefined {
    return this._value;
  }

  isEmpty(): boolean {
    return this._value === undefined;
  }

  equals(other: Impression): boolean {
    return this._value === other._value;
  }
}

File: backend-go/internal/domain/model/impression.go
package model

import (
	"strings"
	"unicode/utf8"
)

const impressionMaxLength = 191

// runeカウントで191文字制限。nilまたは空文字はempty扱い。
type Impression struct {
	value string
	empty bool
}

func NewImpression(value *string) (Impression, error) {
	if value == nil {
		return Impression{empty: true}, nil
	}
	trimmed := strings.TrimSpace(*value)
	if trimmed == "" {
		return Impression{empty: true}, nil
	}
	if utf8.RuneCountInString(trimmed) > impressionMaxLength {
		return Impression{}, NewError(
			"INVALID_IMPRESSION",
			"Impression must be 191 characters or less",
		)
	}
	return Impression{value: trimmed}, nil
}

func (i Impression) Value() *string {
	if i.empty {
		return nil
	}
	value := i.value
	return &value
}

func (i Impression) IsEmpty() bool {
	return i.empty
}

func (i Impression) Equals(other Impression) bool {
	return i == other
}
